Index: pjsip/sources/pjnath/src/pjnath/stun_transaction.c
===================================================================
--- pjsip.orig/sources/pjnath/src/pjnath/stun_transaction.c
+++ pjsip/sources/pjnath/src/pjnath/stun_transaction.c
@@ -398,10 +398,10 @@ PJ_DEF(pj_status_t) pj_stun_client_tsx_r
 	return PJ_SUCCESS;
     }
 
-    if (mod_count) {
+    //if (mod_count) {
         pj_timer_heap_cancel_if_active(tsx->timer_heap, &tsx->retransmit_timer,
                                        TIMER_INACTIVE);
-    }
+    //}
 
     return tsx_transmit_msg(tsx, mod_count);
 }
