Index: pjsip/sources/pjlib/src/pj/ssl_sock_ossl.c
===================================================================
--- pjsip.orig/sources/pjlib/src/pj/ssl_sock_ossl.c
+++ pjsip/sources/pjlib/src/pj/ssl_sock_ossl.c
@@ -522,10 +522,11 @@ static pj_status_t create_ssl(pj_ssl_soc
     init_openssl();
 
     if (ssock->param.proto == PJ_SSL_SOCK_PROTO_DEFAULT)
-	ssock->param.proto = PJ_SSL_SOCK_PROTO_SSL23;
+	ssock->param.proto = PJ_SSL_SOCK_PROTO_TLS1;
 
     /* Determine SSL method to use */
     switch (ssock->param.proto) {
+    case PJ_SSL_SOCK_PROTO_DEFAULT:
     case PJ_SSL_SOCK_PROTO_TLS1:
 	ssl_method = (SSL_METHOD*)TLSv1_method();
 	break;
