Index: pjsip/sources/pjlib/src/pj/os_core_unix.c
===================================================================
--- pjsip.orig/sources/pjlib/src/pj/os_core_unix.c	2014-11-12 18:14:29.151067979 +0100
+++ pjsip/sources/pjlib/src/pj/os_core_unix.c	2014-11-12 20:13:27.154188767 +0100
@@ -49,24 +49,6 @@
 #define SIGNATURE1  0xDEAFBEEF
 #define SIGNATURE2  0xDEADC0DE
 
-#ifndef PJ_JNI_HAS_JNI_ONLOAD
-#  define PJ_JNI_HAS_JNI_ONLOAD    PJ_ANDROID
-#endif
-
-#if defined(PJ_JNI_HAS_JNI_ONLOAD) && PJ_JNI_HAS_JNI_ONLOAD != 0
-
-#include <jni.h>
-
-JavaVM *pj_jni_jvm = NULL;
-
-JNIEXPORT jint JNI_OnLoad(JavaVM *vm, void *reserved)
-{
-    pj_jni_jvm = vm;
-    
-    return JNI_VERSION_1_4;
-}
-#endif
-
 struct pj_thread_t
 {
     char	    obj_name[PJ_MAX_OBJ_NAME];
Index: pjsip/sources/pjmedia/src/pjmedia-audiodev/android_jni_dev.c
===================================================================
--- pjsip.orig/sources/pjmedia/src/pjmedia-audiodev/android_jni_dev.c	2014-11-12 18:03:53.351056558 +0100
+++ pjsip/sources/pjmedia/src/pjmedia-audiodev/android_jni_dev.c	2014-11-12 20:18:10.102193850 +0100
@@ -141,14 +141,21 @@
     &strm_destroy
 };
 
-extern JavaVM *pj_jni_jvm;
+JavaVM *android_jvm;
+
+JNIEXPORT jint JNI_OnLoad(JavaVM *vm, void *reserved)
+{
+    android_jvm = vm;
+
+    return JNI_VERSION_1_4;
+}
 
 static pj_bool_t attach_jvm(JNIEnv **jni_env)
 {
-    if ((*pj_jni_jvm)->GetEnv(pj_jni_jvm, (void **)jni_env,
+    if ((*android_jvm)->GetEnv(android_jvm, (void **)jni_env,
                                JNI_VERSION_1_4) < 0)
     {
-        if ((*pj_jni_jvm)->AttachCurrentThread(pj_jni_jvm, jni_env, NULL) < 0)
+        if ((*android_jvm)->AttachCurrentThread(android_jvm, jni_env, NULL) < 0)
         {
             jni_env = NULL;
             return PJ_FALSE;
@@ -161,7 +168,7 @@
 
 #define detach_jvm(attached) \
     if (attached) \
-        (*pj_jni_jvm)->DetachCurrentThread(pj_jni_jvm);
+        (*android_jvm)->DetachCurrentThread(android_jvm);
 
 /* Thread priority utils */
 /* TODO : port it to pj_thread functions */
