Index: pjproject/_source/pjmedia/src/pjmedia-audiodev/coreaudio_dev.m
===================================================================
--- pjproject.orig/_source/pjmedia/src/pjmedia-audiodev/coreaudio_dev.m
+++ pjproject/_source/pjmedia/src/pjmedia-audiodev/coreaudio_dev.m
@@ -335,6 +335,7 @@ static pj_status_t ca_factory_init(pjmed
     }
 #endif
 
+#if COREAUDIO_DO_NOT_SET_CATEGORY != 1
     /* Initialize audio session category and mode */
     {
 	AVAudioSession *sess = [AVAudioSession sharedInstance];
@@ -352,6 +353,9 @@ static pj_status_t ca_factory_init(pjmed
 	if (err) {
             PJ_LOG(3, (THIS_FILE,
    	               "Warning: failed settting audio session category"));
+	} else {
+    	   PJ_LOG(5, (THIS_FILE,
+   	               "Audio category set to PlayAndRecord."));
 	}
 
 	if ([sess respondsToSelector:@selector(setMode:error:)] &&
@@ -360,6 +364,7 @@ static pj_status_t ca_factory_init(pjmed
 	    PJ_LOG(3, (THIS_FILE, "Warning: failed settting audio mode"));
 	}
     }
+#endif
 
     cf_instance = cf;
 #endif
