Index: pjproject/sources/pjnath/src/pjnath/turn_sock.c
===================================================================
--- pjproject.orig/sources/pjnath/src/pjnath/turn_sock.c
+++ pjproject/sources/pjnath/src/pjnath/turn_sock.c
@@ -675,8 +675,20 @@ static pj_status_t turn_on_send_pkt(pj_t
 
     status = pj_activesock_send(turn_sock->active_sock, &turn_sock->send_key,
 				pkt, &len, 0);
+
+    // Is status indicating Destination required iOS bug? Error EDESTADDRREQ, Destination address required.
+    if (status == PJ_STATUS_FROM_OS(39)){
+        show_err(turn_sock, "socket send()", status);
+        PJ_LOG(4, (turn_sock->obj_name, "Socket does not work, status %d. Trying with destination address", status));
+
+        // On fail try to send with destination, may be iOS bug with closed socket.
+        status = pj_activesock_sendto(turn_sock->active_sock, &turn_sock->send_key,
+                                      pkt, &len, 0, dst_addr, dst_addr_len);
+    }
+
     if (status != PJ_SUCCESS && status != PJ_EPENDING) {
 	show_err(turn_sock, "socket send()", status);
+        PJ_LOG(4, (turn_sock->obj_name, "Socket does not work, status %d", status));
     }
 
     return status;
